client/state/data-layer/http/index.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * External dependencies
+ */
+import superagent from 'superagent';
+
+/***
+ * Internal dependencies
+ */
+import { extendAction } from 'state/utils';
+import { HTTP_REQUEST } from 'state/action-types';
+import { failureMeta, successMeta } from 'state/data-layer/wpcom-http';
+
+/***
+ * Handler to perorm an http request based on `HTTP_REQUEST` action parameters:
+ * {String} url the url to request
+ * {String} method the method we should use in the request: GET, POST etc.
+ * {Array<{Object}>} headers array of { key: '', value: '' } pairs for the request headers
+ * {Object} body data send as body
+ * {Boolean} withCredentials save cookie set on request
+ * {Action} onSuccess action to dispatch on success with data meta
+ * {Action} onFailure action to dispatch on failure with error meta
+ *
+ * @param {Function} dispatch redux store dispatch
+ * @param {Function} action dispathed action we need to handle
+ * @returns {Promise} promise of the handled request
+ */
+const httpHandler = ( { dispatch }, action ) => {
+	const {
+		url,
+		method,
+		headers,
+		body,
+		withCredentials,
+		onSuccess,
+		onFailure
+	} = action;
+
+	let request = superagent( method, url );
+
+	if ( withCredentials ) {
+		request = request.withCredentials();
+	}
+
+	if ( Array.isArray( headers ) ) {
+		headers.forEach( header => {
+			if ( header.key && header.value ) {
+				request = request.set( header.key, header.value );
+			}
+		} );
+	}
```

Comment:
if we default `headers = []` then we can rest in the safety of always having an array. `forEach` on an empty list is a noop which means it's still safe.

further, if we're worried about not setting the appropriate headers (having `key` and `value`) maybe we could validate those in a different place than when transforming them. I'm assuming that any request with invalid headers should actually fail loudly and quickly whereas this code will hide them.

```js
if ( ! validate( headers ) ) {
	throw new TypeError( 'Invalid headers supplied: headers require a key and value property, found ' + headers );
}

request = headers.reduce( header => request.set( header.key, header.value ), request );
```

note also that I think the operation we are looking for is a _reduction_ instead of a _forEach_ since we are iteratively transforming some value and using the result of that transformation.